### ISSUE TITLE
(check-false (json-equal? (string->jsexpr "1.000000000000000000000001") 1.0)) fails

### ISSUE THREAD

**Issue by tonyg** [jessealama/argo#5]
JSON doesn't assign any meaning to its `number` production. JSON Schema does: it requires that `number`s denote decimal numbers.

Racket's `string->jsexpr` represents non-integer JSON `number`s as `float`s, meaning that it cannot exactly represent certain JSON numbers. Implementing a JSON Schema `json-equals?` in terms of `string->jsexpr` can therefore not work.

The problem arises with non-integer cases like `(string->jsexpr "1.000000000000000000000001")` as well as with integer cases like `(string->jsexpr "1e10000")`.

**Comment by samth** [comment]
The easiest fix is probably just to read numbers as exact.

**Comment by tonyg** [comment]
Nice: parameter [read-decimal-as-inexact](https://docs.racket-lang.org/reference/Reading.html#%28def._%28%28quote._~23~25kernel%29._read-decimal-as-inexact%29%29)

**Comment by jessealama** [comment]
Am I missing something?

    (parameterize ([read-decimal-as-inexact #f])
      (= 1.000000000000000000000000000001 1.0))

yields (unexpectedly) `#t`.

**Comment by tonyg** [comment]
It has already been read before it starts being evaluated. Try:

```racket
(parameterize ([read-decimal-as-inexact #f])
  (= (read (open-input-string "1.000000000000000000000000000001")) 1.0))
```

**Comment by jessealama** [comment]
D'oh, you're right, of course. Thanks for the tip!

To properly address the issue, I'll need to extend the class of jsexpr values to include rationals. (At the moment, rationals are banned. Only finite inexact numbers and exact integers are legal.)

**Comment by jessealama** [comment]
I've now defined an extension of the the jsexpr? values so that rationals are allowed. Looking at the problem description again, and equipped with the new "rational JSON", I wonder what the real issue is.

I think that the problem can't really be said to lie with jsexpr, because it's arguably doing the right thing according to the JSON specs. The `string->jsexpr` bit is where things go awry; Argo (more precisely, `json-equal?` then gets the "wrong" inputs but arguably does the right thing with them. Given this, I feel stuck, because I don't know which library to blame.

To remedy the conundrum, I propose reformulating the issue as a bug with JSON Schema validation (and thus something that properly belongs to Argo). The bug goes something like this: the schema:

    { "const": 1.000000000000000000001 }

falsely admits the instance 1.0. (This is a tiny schema whose "instance space" has cardinality 1.)

This can be reproduced by saving that schema and that number to two files and running Argo on them. It can reproduced in a REPL like this:

    > (require argo)
    > (adheres-to-schema? 1.0 (hasheq 'const 1.00000000000000000000000000001))
    #t

What do you think?

**Comment by jessealama** [comment]
If I may put it differently: where's the bug in *Argo* here?

One way of reading the discussion so far is to say that the real culprit is in the definition of `jsexpr?`, because there rational numbers are ruled out:

    > (require json)
    > (jsexpr? #e1.000000000000000000000000000000001)
    #f
    > (jsexpr? 1.000000000000000000000000000000001)
    #t

**Comment by samth** [comment]
The problem with changing `jsexpr?` is that then you can't round-trip data through JSON accurately, because JSON (as a grammar) can't distinguish between `#e1.5` and `#i1.5`. `string->jsexpr` could be changed to always produce exact rationals, but that's very unlikely to be what most people want.

**Comment by tonyg** [comment]
Yes, what Sam said. It's a problem with trying to use a JSON reader as if it were strict enough to satisfy the demands of JSON Schema.

One thing you could try is exporting some kind of `string->json-schema-instance` (and similar routines) which uses `string->jsexpr` internally while manipulating `read-decimal-as-inexact`. Then document them as being like the routines exported from `json`, but stricter, as befits the requirements of JSON Schema.

For `(adheres-to-schema? 1.0 (hasheq 'const 1.00000000000000000000000000001))` I think you're getting what you ask for: by the time anything JSON or JSON-Schema specific sees the values, they've already been read by the generic Racket reader. However, you could expect that `(adheres-to-schema? 1.0 (string->json-schema-instance "{ \"const\": 1.00000000000000000000000000001 }"))` would yield `#f`, and that `(adheres-to-schema? 1.0 (string->json-schema-instance "{ \"const\": 1e0 }"))` would yield `#t`.

Put another way: plain old JSON doesn't mean anything, so you can't use even a 100% spec conformant JSON reader and expect it to be compatible with JSON Schema's notion of instance equality. You have to vet each one individually.

**Comment by jackfirth** [comment]
Here's what the relevant `application/json` specification text has to say about numeric precision limits in [RFC 8259 Section 6](https://tools.ietf.org/html/rfc8259#section-6):

> This specification allows implementations to set limits on the range and precision of numbers accepted.  Since software that implements IEEE 754 binary64 (double precision) numbers [IEEE754] is generally available and widely used, good interoperability can be achieved by implementations that expect no more precision or range than these provide, in the sense that implementations will approximate JSON numbers within the expected precision.  A JSON number such as 1E400 or 3.141592653589793238462643383279 may indicate potential interoperability problems, since it suggests that the software that created it expects receiving software to have greater capabilities for numeric magnitude and precision than is widely available.
>
> Note that when such software is used, numbers that are integers and are in the range `[-(2**53)+1, (2**53)-1]` are interoperable in the sense that implementations will agree exactly on their numeric values.

The JSON Schema draft spec has this to say about numeric equality in [draft-handrews-json-schema-01 Section 4.2.3](https://tools.ietf.org/html/draft-handrews-json-schema-01#section-4.2.3) (emphasis mine):

> Two JSON instances are said to be equal if and only if they are of the same type and have the same value according to the data model. Specifically, this means:
>
> - both are null; or
> - both are true; or
> - both are false; or
> - both are strings, and are the same codepoint-for-codepoint; or
> - both are numbers, and have the **same mathematical value**; or
> - both are arrays, and have an equal value item-for-item; or
> - both are objects, and each property in one has exactly one property with a key equal to the other's, and that other property has an equal value.
>
> Implied in this definition is that arrays must be the same length, objects must have the same number of members, properties in objects are unordered, there is no way to define multiple properties with the same key, and mere formatting differences (indentation, placement of
commas, trailing zeros) are insignificant.

Note the use of the term *mathematical value*; that term is not defined in the JSON spec or the JSON Schema draft spec. In fact, neither spec makes any other mention at all of that term. Sigh.

**Comment by tonyg** [comment]
@jackfirth, JSON Schema doesn't define "mathematical value", but it *does* (in section 4.2.1) define what a "number" is:

>  number:  An arbitrary-precision, base-10 decimal number value, **from the JSON "number" production**

(Emphasis mine.) I read this as making a connection between the syntax of a document and its denotation. I also read it as suggesting a reasonable interpretation for "mathematical value".

Later in the same section, we have:

>  Whitespace and formatting concerns, including different lexical representations of numbers that are equal within the data model, are thus outside the scope of JSON Schema.

...which seems to me to be saying that documents containing "1", "1.0", "1e0", "0.001e3" and "100e-2" should all be treated alike from the perspective of JSON Schema; that JSON Schema assumes some mapping from document text to JSON Schema instance that reflects JSON Schema's definition of "number" and its definition of instance equality.

**Comment by jessealama** [comment]
Proof of life/status report:

I'm about to push a fix for this problem. To address the issue, I adopted Sam's suggestion.

At first, my changes were contained entirely within Argo. But I realized that I was writing essentially my own stand-alone JSON library. I separated my work from Argo and packaged it up as a stand-alone Racket package (https://github.com/jessealama/ejs) that should take care of this problem. It's essentially the same as the built-in JSON library, but with one twist: numbers are exact reals.

The next task is to switch out the built-in JSON library with my homemade one. At that point, this ticket can be closed.
